File: examples/async-await/async-await.kt
import kotlinx.coroutines.*
import kotlin.system.*

suspend fun sumParallel(): Int = coroutineScope {
    val a = async { f(100, 1) }
    val b = async { f(200, 5) }
    a.await() + b.await()
}

suspend fun f(v: Int, s: Int): Int {
    println("Fetching $v...")
    delay(s * 1000L)
    return v
}

fun main() = runBlocking {
    val result = sumParallel()
    println(result)
}

File: examples/async-await/async-await.ts
async function sumParallel() {
  const aPromise = f(100, 1);
  const bPromise = f(200, 5);
  const [a, b] = await Promise.all([
    aPromise,
    bPromise,
  ]);
  return a + b;
}

async function f(v: number, s: number) {
  console.log(`Fetching ${v}...`);
  // delay
  await new Promise((resolve) => {
    setTimeout(resolve, s * 1000);
  });
  return v;
}

sumParallel().then(console.log);
